File: examples/CSharpProd/JsonConfig/JsonConfigExample.cs
using System;
using System.Data;
using System.Linq;
using System.Net.Http;

using System.Threading.Tasks;

using Microsoft.Extensions.Configuration;
using Newtonsoft.Json;

using NBomber;
using NBomber.Contracts;
using NBomber.Contracts.Stats;
using NBomber.CSharp;
using NBomber.Plugins.Network.Ping;

using JsonSerializer = System.Text.Json.JsonSerializer;

namespace CSharpProd.JsonConfig
{
    // in this example we use JSON configuration files:
    // - 'config.json' to configure load test settings
    // - 'infra-config.json' to configure infrastructure settings (logging, worker plugins, reposting sinks)
    // - 'user-feed.json' to configure data source to feed from

    public class User
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    public class UserResponse
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string Email { get; set; }
        public string Phone { get; set; }
    }

    public class PostResponse
    {
        public string Id { get; set; }
        public string UserId { get; set; }
        public string Title { get; set; }
        public string Body { get; set; }
    }

    public class CustomScenarioSettings
    {
        public string BaseUri { get; set; }
    }

    public class CustomPluginSettings
    {
        public string Message { get; set; }
    }

    public class CustomReportingSinkSettings
    {
        public string Message { get; set; }
    }

    public class CustomPlugin : IWorkerPlugin
    {
        private readonly DataSet _pluginStats = new DataSet();
        private CustomPluginSettings _customPluginSettings;

        public CustomPlugin(CustomPluginSettings customPluginSettings) =>
            _customPluginSettings = customPluginSettings;

        public string PluginName => nameof(CustomPlugin);

        public Task Init(IBaseContext context, IConfiguration infraConfig)
        {
            var logger = context.Logger.ForContext<CustomPlugin>();

            _customPluginSettings =
                infraConfig.GetSection(nameof(CustomReportingSink)).Get<CustomPluginSettings>()
                ?? _customPluginSettings;

            var settingsJson = JsonSerializer.Serialize(_customPluginSettings);

            logger.Information($"{nameof(CustomPlugin)} settings: {settingsJson}");

            return Task.CompletedTask;
        }

        public Task Start() => Task.CompletedTask;

        public Task<DataSet> GetStats(OperationType currentOperation) => Task.FromResult(_pluginStats);

        public string[] GetHints() => Array.Empty<string>();

        public Task Stop() => Task.CompletedTask;

        public void Dispose()
        {
        }
    }

    public class CustomReportingSink : IReportingSink
    {
        private CustomReportingSinkSettings _customReportingSinkSettings;

        public CustomReportingSink(CustomReportingSinkSettings customReportingSinkSettings) =>
        _customReportingSinkSettings = customReportingSinkSettings;

        public string SinkName => nameof(CustomReportingSink);

        public Task Init(IBaseContext context, IConfiguration infraConfig)
        {
            var logger = context.Logger.ForContext<CustomReportingSink>();

            _customReportingSinkSettings =
                infraConfig.GetSection(nameof(CustomReportingSink)).Get<CustomReportingSinkSettings>()
                ?? _customReportingSinkSettings;

            var settingsJson = JsonSerializer.Serialize(_customReportingSinkSettings);

            logger.Information($"{nameof(CustomReportingSink)} settings: {settingsJson}");

            return Task.CompletedTask;
        }

        public Task Start() => Task.CompletedTask;
        public Task SaveRealtimeStats(ScenarioStats[] stats) => Task.CompletedTask;
        public Task SaveFinalStats(NodeStats[] stats) => Task.CompletedTask;
        public Task Stop() => Task.CompletedTask;

        public void Dispose()
        {
        }
    }

    public static class JsonConfigExample
    {
        private static CustomScenarioSettings _customSettings = new CustomScenarioSettings();

        private static Task ScenarioInit(IScenarioContext context)
        {
            _customSettings = context.CustomSettings.Get<CustomScenarioSettings>();

            context.Logger.Information($"test init received CustomSettings: {_customSettings}");

            return Task.CompletedTask;
        }

        public static void Run()
        {
            var userFeed = Feed.CreateRandom(
                "userFeed",
                FeedData.FromJson<User>("./JsonConfig/Configs/user-feed.json")
            );

            var httpFactory = ClientFactory.Create("http_factory",
                initClient: (number,context) => Task.FromResult(new HttpClient { BaseAddress = new Uri(_customSettings.BaseUri)}));

            var getUser = Step.Create("get_user", httpFactory, userFeed, async context =>
            {
                var userId = context.FeedItem;
                var url = $"users?id={userId}";

                var response = await context.Client.GetAsync(url, context.CancellationToken);
                var json = await response.Content.ReadAsStringAsync();

                // parse JSON
                var users = JsonConvert.DeserializeObject<UserResponse[]>(json);

                return users?.Length == 1
                    ? Response.Ok(users.First()) // we pass user object response to the next step
                    : Response.Fail("not found user");
            });

            // this 'getPosts' will be executed only if 'getUser' finished OK.
            var getPosts = Step.Create("get_posts", httpFactory, async context =>
            {
                var user = context.GetPreviousStepResponse<UserResponse>();
                var url = $"posts?userId={user.Id}";

                var response = await context.Client.GetAsync(url, context.CancellationToken);
                var json = await response.Content.ReadAsStringAsync();

                // parse JSON
                var posts = JsonConvert.DeserializeObject<PostResponse[]>(json);

                return posts?.Length > 0
                    ? Response.Ok()
                    : Response.Fail();
            });

            var scenario1 = ScenarioBuilder
                .CreateScenario("rest_api", getUser, getPosts)
                .WithInit(ScenarioInit);

            // the following scenario will be ignored since it is not included in target scenarios list of config.json
            var scenario2 = ScenarioBuilder
                .CreateScenario("ignored", getUser, getPosts);

            // settings for plugins and reporting sinks are overriden in infra-config.json
            var pingPlugin = new PingPlugin();

            var customPlugin = new CustomPlugin(new CustomPluginSettings
                {Message = "Plugin is configured via constructor"});

            var customReportingSink = new CustomReportingSink(new CustomReportingSinkSettings
                {Message = "Reporting sink is configured via constructor"});

            NBomberRunner
                .RegisterScenarios(scenario1, scenario2)
                .WithWorkerPlugins(pingPlugin, customPlugin)
                .WithReportingSinks(customReportingSink)
                .LoadConfig("./JsonConfig/Configs/config.json")
                .LoadInfraConfig("./JsonConfig/Configs/infra-config.json")
                .Run();
        }
    }
}

File: examples/FSharpProd/JsonConfig/JsonConfigExample.fs
module FSharpProd.JsonConfig.JsonConfigExample

open System
open System.Data
open System.Net.Http
open System.Threading.Tasks

open FSharp.Control.Tasks.NonAffine
open Microsoft.Extensions.Configuration
open Newtonsoft.Json

open NBomber
open NBomber.Contracts
open NBomber.FSharp
open NBomber.Plugins.Network.Ping
open NBomber.Sinks.InfluxDB

// in this example we use JSON configuration files:
// - 'config.json' to configure load test settings
// - 'infra-config.json' to configure infrastructure settings (logging, worker plugins, reposting sinks)
// - 'user-feed.json' to configure data source to feed from

module Option =

    let ofRecord (value: 'T) =
        let boxed = box(value)
        if isNull boxed then None
        else Some value

[<CLIMutable>]
type User = {
    Id: int
    Name: string
}

[<CLIMutable>]
type UserResponse = {
    Id: string
    Name: string
    Email: string
    Phone: string
}

[<CLIMutable>]
type PostResponse = {
    Id: string
    UserId: string
    Title: string
    Body: string
}

[<CLIMutable>]
type CustomScenarioSettings = {
    BaseUri: string
}

[<CLIMutable>]
type CustomPluginSettings = {
    Message: string
}

[<CLIMutable>]
type CustomReportingSinkSettings = {
    Message: string
}

type CustomPlugin(customPluginSettings: CustomPluginSettings) =
    let mutable _pluginStats = new DataSet()

    interface IWorkerPlugin with
        member _.PluginName = nameof(CustomPlugin)

        member _.Init(context, infraConfig) =
            let logger = context.Logger.ForContext<CustomPluginSettings>()

            let settings =
                infraConfig.GetSection(nameof(CustomPlugin)).Get<CustomPluginSettings>()
                |> Option.ofRecord
                |> Option.defaultValue customPluginSettings

            logger.Information($"{nameof(CustomPlugin)} settings: {settings}")

            Task.CompletedTask

        member _.Start() = Task.CompletedTask
        member _.GetStats(currentOperation) = Task.FromResult(_pluginStats)
        member _.GetHints() = Array.empty
        member _.Stop() = Task.CompletedTask
        member _.Dispose() = ()

type CustomReportingSink(customReportingSinkSettings: CustomReportingSinkSettings) =
    interface IReportingSink with
        member _.SinkName = nameof(CustomReportingSink)
        member _.Init(context, infraConfig) =
            let logger = context.Logger.ForContext<CustomReportingSinkSettings>()

            let settings =
                infraConfig.GetSection(nameof(CustomReportingSink)).Get<CustomReportingSinkSettings>()
                |> Option.ofRecord
                |> Option.defaultValue customReportingSinkSettings

            logger.Information($"{nameof(CustomReportingSink)} settings: {settings}")

            Task.CompletedTask
        member _.Start() = Task.CompletedTask
        member _.SaveRealtimeStats(stats) = Task.CompletedTask
        member _.SaveFinalStats(stats) = Task.CompletedTask
        member _.Stop() = Task.CompletedTask
        member _.Dispose() = ()

let run () =

    let mutable _customSettings = Unchecked.defaultof<CustomScenarioSettings>

    let scenarioInit (context: IScenarioContext) = task {
        _customSettings <- context.CustomSettings.Get<CustomScenarioSettings>()

        context.Logger.Information($"test init received CustomSettings: {_customSettings}")
    }

    let userFeed = FeedData.fromJson<User> "./JsonConfig/Configs/user-feed.json"
                   |> Feed.createRandom "userFeed"

    let httpFactory =
        ClientFactory.create(name = "http_factory",
                             initClient = fun (number,context) -> task {
                                 return new HttpClient(BaseAddress = Uri(_customSettings.BaseUri))
                             })

    let getUser = Step.create("get_user",
                              clientFactory = httpFactory,
                              feed = userFeed,
                              execute = fun context -> task {

        let userId = context.FeedItem
        let url = $"users?id={userId.Id}"

        let! response = context.Client.GetAsync(url, context.CancellationToken)
        let! json = response.Content.ReadAsStringAsync()

        // parse JSON
        let users = json
                    |> JsonConvert.DeserializeObject<UserResponse[]>
                    |> ValueOption.ofObj

        match users with
        | ValueSome usr when usr.Length = 1 ->
            return Response.ok(usr.[0]) // we pass user object response to the next step

        | _ -> return Response.fail($"not found user: {userId.Id}")
    })

    // this 'getPosts' will be executed only if 'getUser' finished OK.
    let getPosts = Step.create("get_posts",
                               clientFactory = httpFactory,
                               execute = fun context -> task {

        let user = context.GetPreviousStepResponse<UserResponse>()
        let url = $"posts?userId={user.Id}"

        let! response = context.Client.GetAsync(url, context.CancellationToken)
        let! json = response.Content.ReadAsStringAsync()

        // parse JSON
        let posts = json
                    |> JsonConvert.DeserializeObject<PostResponse[]>
                    |> ValueOption.ofObj

        match posts with
        | ValueSome ps when ps.Length > 0 ->
            return Response.ok()

        | _ -> return Response.fail()
    })

    let scenario1 =
        Scenario.create "rest_api" [getUser; getPosts]
        |> Scenario.withInit scenarioInit

    // the following scenario will be ignored since it is not included in target scenarios list of config.json
    let scenario2 = Scenario.create "ignored" [getUser]

    // settings for plugins and reporting sinks are overriden in infra-config.json
    let pingPlugin = new PingPlugin()
    let customPlugin = new CustomPlugin({ Message = "Plugin is configured via constructor" })
    let influxDbReportingSink = new InfluxDBSink();
    let customReportingSink = new CustomReportingSink({ Message = "Reporting sink is configured via constructor" })

    NBomberRunner.registerScenarios [scenario1; scenario2]
    |> NBomberRunner.withWorkerPlugins [pingPlugin; customPlugin]
    |> NBomberRunner.withReportingSinks [influxDbReportingSink; customReportingSink]
    |> NBomberRunner.loadConfig "./JsonConfig/Configs/config.json"
    |> NBomberRunner.loadInfraConfig "./JsonConfig/Configs/infra-config.json"
    |> NBomberRunner.run
    |> ignore
